### ISSUE TITLE
Add a lint for `defer recover()`

### ISSUE THREAD

**Issue by CAFxX** [golang/lint#469]
The following [is allowed](https://github.com/golang/go/issues/142730), but it doesn't do anything useful and especially not what one could expect ("recover and ignore the panic") at a cursory reading. I'd argue there's no good reason why it should ever be written:

```
func foo() {
  defer recover()
  // ...
}
```

it should instead be:

```
func foo() {
  defer func() {
    recover()
  }()
  // ...
}
```

lint should probably emit a warning if it encounters `defer recover()`.

While there's an argument to be made that possibly recovering and ignoring the panic is a questionable practice, it does not change the fact that the specific form discussed above should generate a warning of some sort.

A quick search reveals that this actually happens in the wild, although it's not exceedingly common: https://www.google.com/search?q=intext:"defer+recover()"+ext:go+site:github.com

Not sure if this should go in vet or lint. I can move it if needed.

**Comment by elgs** [comment]
Why? I thought it's perfectly fine to write it either way.

**Comment by neild** [comment]
It should also be possible to check for some cases where `recover` is called in a function which is not itself deferred:

```go
func f1() {
  recover()
}

func f2() {
  defer f1() // ok
  f1()       // not ok: non-defer call to recovering function
}
```

I suppose you could construct a scenario where a function wants to recover if it's deferred, but not otherwise, but I can't think of a non-contrived reason to do that.

**Comment by mvdan** [comment]
Thank you for submitting this issue! As per https://github.com/golang/go/issues/38968, we are freezing and deprecating golint. There's no drop-in replacement to golint per se, but you should find that [Staticcheck](https://staticcheck.io/) works well in encouraging good Go code, much like golint did in the past, since it also includes style checks. There's always `gofmt` and `go vet` too, of course.

If you would like to contribute further, I'd encourage you to engage [Staticcheck's issue tracker](https://github.com/dominikh/go-tools/issues) or look at [vet's open issues](https://github.com/golang/go/issues?q=is%3Aissue+is%3Aopen+cmd%2Fvet+in%3Atitle), as they are both actively maintained. If you have an idea that doesn't fit into either of those tools, you could look at [other Go linters](https://github.com/golangci/awesome-go-linters), or write your own - these days it's fairly straightforward with [go/analysis](https://pkg.go.dev/golang.org/x/tools/go/analysis).

To help avoid confusion, I'm closing all issues before we freeze the repository. If you have any feedback, you can leave a comment on the proposal thread where it was decided to deprecate golint - though note that the proposal has been accepted for nearly a year. Thanks!
